### ISSUE TITLE
Docs/Setup: Reduce compile times with "Build Phase" script change.

### ISSUE THREAD

**Issue by matt-holden** [apollographql/apollo-ios#147]
I recently had a PR merged into an open-source iOS app project that uses Apollo: https://github.com/rnystrom/GitHawk/pull/327

You can see a more detailed explanation in that PR summary – in short, overwriting `API.swift` on every compile, even if its output was unchanged from the last build, was triggering a *massive* increase in compile times.

The project had been using the exact build phase script from: http://dev.apollodata.com/ios/installation.html#adding-build-step

The following change prevents the compiler from needing to rebuild `API.swift`, and all swift files that reference it, when it hasn't changed. _This removed a full 60 seconds from the app's compile time._

```sh
APOLLO_FRAMEWORK_PATH="$(eval find $FRAMEWORK_SEARCH_PATHS -name "Apollo.framework" -maxdepth 1)"

if [ -z "$APOLLO_FRAMEWORK_PATH" ]; then
  echo "error: Couldn't find Apollo.framework in FRAMEWORK_SEARCH_PATHS; make sure to add the framework to your project."
  exit 1
fi

cd "${SRCROOT}/${TARGET_NAME}"

##############################
# Modifications to original script below:
##############################

TEMP_FILE=$(mktemp)
[[ -f API.swift ]] || touch API.swift

$APOLLO_FRAMEWORK_PATH/check-and-run-apollo-codegen.sh generate $(find . -name '*.graphql') --schema schema.json --output $TEMP_FILE

cmp API.swift $TEMP_FILE || cp $TEMP_FILE API.swift
```

Cheers!

**Comment by martijnwalraven** [comment]
Ah, that's a neat trick, thanks! I've heard more reports from people about increases in compile times. One of the solutions might be to generate multiple smaller files instead of one big `API.swift`, but that has its own complications. So I'll make sure to mention this in an update to the docs, because it may help a lot of people.

**Comment by MrAlek** [comment]
There is actually a built in way to do this in Xcode – using input & output files. (http://indiestack.com/2014/12/speeding-up-custom-script-phases/)

Here's an example of how we're using it with SwiftGen:

![skarmavbild 2017-09-21 kl 14 25 18](https://user-images.githubusercontent.com/1496135/30695558-d005222a-9ed8-11e7-8f90-cddcb1165e33.png)

In the case of Apollo Codegen, what you would do is to specify all your `.graphql` files + the `schema.json` file as input files, and then your `API.swift` as an output file.

That way, Xcode doesn't run the script unless either any of the input files have changed, or if the `API.swift` file is missing.

**Comment by matt-holden** [comment]
@MrAlek ^ by far the most useful thing I've learned this week.

**Comment by martijnwalraven** [comment]
@MrAlek: It would be even nicer if we could get this working with a build rule, so you only have to add the `.graphql` files to your project as sources. That should also help with generating a `.swift` file for every `.graphql` file (fixing #151) without forcing you to add the Swift sources to your project manually. It's not entirely clear to me how that would work. Have you ever used build rules? I found an example [here](http://ootips.org/yonat/auto-compile-proto-files/).

**Comment by MrAlek** [comment]
@martijnwalraven I haven't used build rules before but it looks promising. I wonder how it would work with indexing for code completion, documentation, etc. Also if Xcode is smart enough to only run the build when the `.graphl` files change.

**Comment by martijnwalraven** [comment]
I've added experimental support for generating multiple files to `apollo-codegen`, available in Apollo iOS 0.7.0-alpha.7. 

If you specify a directory instead of a file as `--output`, it will generate matching `.graphql.swift` files for every `.graphql` file. In addition, it will generate `Types.graphql.swift` with shared types used in the input files.

Would be great to get feedback and see if this helps improve compilation speed. 

I would still like to find a better way to integrate this with Xcode, so you don't have to manually add `.graphql.swift` files to your project. 

I've tried to use Xcode build rules, but there are some rough edges. Any help in getting this to work is highly appreciated!

To try it out, you'll have to add your `.graphql` files under 'Compile Sources' and add the following build rule:

<img width="1975" alt="screen shot 2017-09-30 at 15 01 02" src="https://user-images.githubusercontent.com/54485/31046226-7b335892-a5f5-11e7-93ca-eedf4666b47c.png">

One downside of build rules is that they work on individual files. Because `.graphql` files have dependencies (on fragments defined in other files for example), we always need all files as input to parsing and validation. I've added an `--only <filePath>` option that only generates output for the specified file. That works, but I'm having trouble getting the build system to automatically recognize `Types.graphql.swift`. I can't specify it as an output file, because that adds it to the compile set multiple times, resulting in duplication errors.

**Comment by martijnwalraven** [comment]
I [opened a PR](https://github.com/apollographql/xcode-graphql/pull/6) with a (probably failed) attempt to add an Apollo compiler plugin, which seems like it would have given us the best developer experience. Unfortunately it seems we need to look at other solutions.

**Comment by jgavris** [comment]
One nice way to do this is to create a local podspec for your GraphQL API. Cocoapods recently added automatic generation of input files rules for pods frameworks. 

https://github.com/CocoaPods/CocoaPods/pull/6800

**Comment by parrotbait** [comment]
> I've added experimental support for generating multiple files to `apollo-codegen`, available in Apollo iOS 0.7.0-alpha.7.
> 
> If you specify a directory instead of a file as `--output`, it will generate matching `.graphql.swift` files for every `.graphql` file. In addition, it will generate `Types.graphql.swift` with shared types used in the input files.
> 
> Would be great to get feedback and see if this helps improve compilation speed.
> 
> I would still like to find a better way to integrate this with Xcode, so you don't have to manually add `.graphql.swift` files to your project.
> 
> I've tried to use Xcode build rules, but there are some rough edges. Any help in getting this to work is highly appreciated!
> 
> To try it out, you'll have to add your `.graphql` files under 'Compile Sources' and add the following build rule:
> 
> <img alt="screen shot 2017-09-30 at 15 01 02" width="1975" src="https://user-images.githubusercontent.com/54485/31046226-7b335892-a5f5-11e7-93ca-eedf4666b47c.png">
> 
> One downside of build rules is that they work on individual files. Because `.graphql` files have dependencies (on fragments defined in other files for example), we always need all files as input to parsing and validation. I've added an `--only <filePath>` option that only generates output for the specified file. That works, but I'm having trouble getting the build system to automatically recognize `Types.graphql.swift`. I can't specify it as an output file, because that adds it to the compile set multiple times, resulting in duplication errors.

Thank you so much. I've tested this and it's made a huge difference for my project. A single API.swift file used to take Xcode ~450 seconds. We don't even have that complex setup. I moved codegen away from being an automatic step because it was such a productivity kill.

With multiple files (I now have 34 files instead of 1) and compilation time from clean is down to a much more manageable 200 seconds. My presumption is that it's benefiting from compilation parallelsation in Xcode. I guess a single file cannot be built in parallel. 

It's also a more preferable setup also because I can change queries and mutations without experiencing a recompile of this giant file.  The downside is that all new files need to be added to Xcode manually but that's well worth it for the productivity gains.

Personally I would recommend that this setup should be the default and recommended for all Apollo setups, in tutorials etc.

**Comment by designatednerd** [comment]
I've added an [Advanced Codegen Tips And Tricks section](https://www.apollographql.com/docs/ios/installation/#advanced-codegen-tips-and-tricks) to our installation documentation that outlines workarounds for setting up input and output files as well as having multiple files generated instead of one ginormous one. 

If there are any other things you would like to add, please open a new issue (or even better, a PR!). Thank you!

(Note that I'm still planning to [add XCFileList generation](https://github.com/apollographql/apollo-ios/issues/636) as part of the Swift Codegen project, but that workaround should get you through until that's up and running)
